### PR TITLE
fips support in podvm builds

### PR DESCRIPTION
**Pull request by snir911**
for now tested only with RHEL, some adaptions required for ubuntu support

### CODE REVIEW

File: podvm/Dockerfile.podvm_binaries.rhel [hunk 1]
```diff
@@ -41,6 +41,6 @@ RUN cd cloud-api-adaptor/podvm && \
 RUN tar czvf /podvm-binaries.tar.gz -C /src/cloud-api-adaptor/podvm/files usr/ etc/
 RUN tar czvf /pause-bundle.tar.gz -C /src/cloud-api-adaptor/podvm/files pause_bundle/
 
-FROM registry.access.redhat.com/ubi8/ubi:8.7
```
Comment (bpradipt) [inline podvm/Dockerfile.podvm_binaries.rhel:44]: Can you also add a new commit for CentOS to use centos9 stream and codeready-builder ?

**Review by bpradipt — APPROVED** [review]
/lgtm
Few suggestions to include either in this PR or next
1. Updating CentOS to CentOS 9 stream
2. Adding ENABLE_FIPS in the README

**Comment by bpradipt** [comment]
@snir911 it might be helpful to paste test results with FIPS enabled CVM to be sure nothing broke from a functionality standpoint.

**Comment by mkulke** [comment]
I don't have practical experience with deploying FIPS-enabled applications, but I suspect this has implications beyond the kernel and needs to be enforced in the userland, so this applies to all kinds of staticly linked binaries like CAA, kata-agent and guest-components. AFAIK there are no FIPS certified rust crypto-modules, not sure about go. We can link against fips-mode openssl probably, but we'd have to do this consistently.

Should we limit the scope of the flag (`FIPS_KERNEL`) and maybe abort the image build if `ENABLE_FIPS` is set, but we don't support it for a given distro yet?

IMO, we could be more succinct in naming our flags: e.g. `ENABLE_FIPS=true` and `DISABLE_CLOUD_CONFIG=true` could bw shortened to `FIPS=true` `CLOUD_CONFIG=false`.

**Comment by katexochen** [comment]
>  not sure about go

Go can link against boringSSL for FIPS compliant crypto.

I know some customers/use cases require FIPS, but letting a 'I don't like fips' here anyway. :wink: 

Other things of interest might be the plan to use wireguard for podvm tunnel encryption (https://github.com/confidential-containers/cloud-api-adaptor/issues/688), as wg isn't FIPS compliant. Would likely need replacement with IPSec for FIPS.

File: aws/image/misc-settings.sh [hunk 2]
```diff
@@ -99,6 +99,24 @@ ExecStartPre=
...
+# Add vsphere specific commands to execute on remote
+    case $PODVM_DISTRO in
+    centos)
+        #fallthrough
+        ;&
+    rhel)
+        fips-mode-setup --enable
+        ;;
+    ubuntu)
+        # TODO: configure the bootloader to add kernel options "fips=1"
```
[the remark this answers is not in the document]
Comment (snir911) [reply aws/image/misc-settings.sh:113]: It seems there's a helper tool only for Pro subscription, i'm not sure how handle ATM

**Comment by snir911** [comment]
Yes, kernel wouldn't be enough for a whole FIPS compatibility, maybe it would be better to name the variable FIPS_BOOT to make it more clear, WDYT?
I was aware to the binaries incompatibility (mostly in rust) but i didn't think of the tunneling

**Comment by bpradipt** [comment]
> Yes, kernel wouldn't be enough for a whole FIPS compatibility, maybe it would be better to name the variable FIPS_BOOT to make it more clear, WDYT? I was aware to the binaries incompatibility (mostly in rust) but i didn't think of the tunneling

We don't yet use wg based tunnelling. But it's an important consideration.

File: podvm/Dockerfile.podvm_binaries.rhel (hunk 1 above)
Comment (snir911) [reply podvm/Dockerfile.podvm_binaries.rhel:44]: lets do CentOS in separate i hit weird issue with pkgs namings

**Comment by snir911** [comment]
I add a note in docs and changes the var name to BOOT_FIPS.
I intentionally don't add the variable to dockerfile until we'll have better coverage also for Ubuntu (and binaries?)

File: versions.yaml [hunk 3]
```diff
@@ -14,6 +14,12 @@ cloudimg:
       amd64:
         url: https://cloud.centos.org/centos/8-stream/x86_64/images/CentOS-Stream-GenericCloud-8-20220913.0.x86_64.qcow2
         checksum: "sha256:8717251f8e4d2fe3e5032799caae89358c1ba68d65a16b5128a59ec6003aac1c"
+  rhel:
+    9.2: # dummy link, get trial image from: https://www.redhat.com/en/technologies/linux-platforms/enterprise-linux/server/trial
+      amd64:
+        url: https://access.redhat.com/downloads/content/479/ver=/rhel---9/9.2/x86_64/product-software
+        checksum: "sha256:34ff41b5274692c984e3860b21136af8b6ae502744c6c7578dda82002fba0287"
```
Comment (bpradipt) [inline versions.yaml:21]: nit: may be you want to edit the commit on dummy link and also add any other prereq to get the image like rhel id etc

Comment (snir911) [reply versions.yaml:21]: I'm not sure i follow, if you use trial no prerequisites are needed IIUC

**Review by jensfr — APPROVED** [review]
Hi @snir911, changes look good to me!

**Comment by snir911** [comment]
@mkulke @katexochen are you ok with this? i modified the var name to BOOT_FIPS to make it clear it's fips boot compatibility

**Comment by katexochen** [comment]
> mkulke katexochen are you ok with this? i modified the var name to BOOT_FIPS to make it clear it's fips boot compatibility

Okay for me, thanks for asking. I still don't fully get what the benefit of this is. If your userland isn't FIPS compliant, what do you get from a FIPS compliant kernel? The podvm still isn't compliant (and won't be in the near future). Is there any benefit of partial compliance for customers? Just trying to understand the use of this..

**Comment by mkulke** [comment]
> > mkulke katexochen are you ok with this? i modified the var name to BOOT_FIPS to make it clear it's fips boot compatibility
> 
> Okay for me, thanks for asking. I still don't fully get what the benefit of this is. If your userland isn't FIPS compliant, what do you get from a FIPS compliant kernel? The podvm still isn't compliant (and won't be in the near future). Is there any benefit of partial compliance for customers? Just trying to understand the use of this..

yup, same for me. My concern was mostly the broad claim of `FIPS=true`, I understand sometimes there are box-checking exercises which would mandate a FIPS kernel, even if the full scope of the application is not covered

**Comment by snir911** [comment]
Clearly it requires much more work to have CoCo, as whole, FIPS complaint, however, i think it can still useful for PeerPods usecases
* This will allow mimicking FIPS environment for FIPS complaint workloads for testing, CI etc..
* You can allow partial FIPS compatibility which can be helpful in certain cases
* once we'll be able to provide FIPS compatible binaries we'll be already in a better place, for rust IIUC  openssl, boringssl and aws-lc has some sort of FIPS compatability, it requires some investigation

**Comment by katexochen** [comment]
> This will allow mimicking FIPS environment for FIPS complaint workloads for testing, CI etc..

My concern here is that each of such options adds an additional dimension to our build and test matrix. We have a ton of options how things can be configured, but we don't have test to test them. And as many options aren't tested, things are silently broken and the whole thing is difficult to use. If we would actually want to test all the combinations, our CI hours would explode.

That's also why I really aim to keep the matrix low with the new mkosi images: one distro, ideally only one configuration. However, I think mkosi will also make it much easier to add additional configuration to images out of tree.

>  I understand sometimes there are box-checking exercises which would mandate a FIPS kernel, even if the full scope of the application is not covered

If this is required anyway for the box-checkers, I totally understand.

**Comment by mkulke** [comment]
> That's also why I really aim to keep the matrix low with the new mkosi images: one distro, ideally only one configuration. However, I think mkosi will also make it much easier to add additional configuration to images out of tree.

I want to underline this point, every toggle that we add permutations to the testing burden, which is already pretty large atm (cloud vendor, operating system, ...)

**Comment by snir911** [comment]
ok, thanks for your comments, I'll be closing it for now (at least until we'll know how to adapt the binaries for fips as well)

**Pull request closed without merging**